grails-app/services/org/pih/warehouse/data/ProductPackageService.groovy

```diff
@@ -0,0 +1,52 @@
+/**
+ * Copyright (c) 2012 Partners In Health.  All rights reserved.
+ * The use and distribution terms for this software are covered by the
+ * Eclipse Public License 1.0 (http://opensource.org/licenses/eclipse-1.0.php)
+ * which can be found in the file epl-v10.html at the root of this distribution.
+ * By using this software in any fashion, you are agreeing to be bound by
+ * the terms of this license.
+ * You must not remove this notice, or any other, from this software.
+ **/
+package org.pih.warehouse.data
+
+import grails.gorm.transactions.Transactional
+import grails.validation.ValidationException
+import org.pih.warehouse.core.UnitOfMeasure
+import org.pih.warehouse.product.Product
+import org.pih.warehouse.product.ProductPackage
+import org.pih.warehouse.product.ProductSupplier
+
+@Transactional
+class ProductPackageService {
+
+    def createOrUpdate(Map params) {
+        log.info("params: ${params}")
+        Product product = Product.findByProductCode(params.productCode)
+        ProductSupplier productSupplier = ProductSupplier.findByCode(params.productSupplierCode)
+        UnitOfMeasure unitOfMeasure = UnitOfMeasure.findByCode(params.uomCode)
+        ProductPackage productPackage = findOrCreate(params)
+        productPackage.properties = params
+        if (!productPackage.name) {
+            productPackage.name = "${unitOfMeasure.code}/${params.quantity}"
+        }
+        if (!productPackage.description) {
+            productPackage.description = "${unitOfMeasure.name} of ${params.quantity}"
+        }
+        productPackage.product = product
+        productPackage.productSupplier = productSupplier
+        productPackage.uom = unitOfMeasure
```

Comment:
same as above - we can consider avoiding this boilerplate binding by using dto/command approach.

Reply:
Again, the issue is with the data binding mechanism. In controllers we're always passing in IDs so the data binding layer is able to convert the ID to a domain object. When importing data via CSV/Excel, we're passing more natural identifiers like product codes and location numbers. These values would not be handled by the data binding mechanism, making the command object less useful in that you'd still need to do these lookups somewhere.